src/phoenix/server/api/subscriptions.py

```diff
@@ -63,42 +97,309 @@ async def chat_completion(
+            messages = list(
+                _formatted_messages(
+                    messages=messages,
+                    template_language=template_options.language,
+                    template_variables=template_options.variables,
+                )
+            )
         invocation_parameters = llm_client.construct_invocation_parameters(
             input.invocation_parameters
         )
         async with streaming_llm_span(
             input=input,
             messages=messages,
             invocation_parameters=invocation_parameters,
-            db=info.context.db,
             attributes=attributes,
         ) as span:
             async for chunk in llm_client.chat_completion_create(
                 messages=messages, tools=input.tools or [], **invocation_parameters
             ):
                 span.add_response_chunk(chunk)
                 yield chunk
-        yield span.finished_chat_completion
-        info.context.event_queue.put(SpanInsertEvent(ids=(span.project_id,)))
+        if span.error_message is not None:
+            yield ChatCompletionSubscriptionError(message=span.error_message)
+        async with info.context.db() as session:
+            if (
+                playground_project_id := await session.scalar(
+                    select(models.Project.id).where(models.Project.name == PLAYGROUND_PROJECT_NAME)
+                )
+            ) is None:
+                playground_project_id = await session.scalar(
+                    insert(models.Project)
+                    .returning(models.Project.id)
+                    .values(
+                        name=PLAYGROUND_PROJECT_NAME,
+                        description="Traces from prompt playground",
+                    )
+                )
+            db_span = span.add_to_session(session, playground_project_id)
+            await session.flush()
+            yield FinishedChatCompletion(span=to_gql_span(db_span))
+        info.context.event_queue.put(SpanInsertEvent(ids=(playground_project_id,)))
+
+    @strawberry.subscription
+    async def chat_completion_over_dataset(
+        self, info: Info[Context, None], input: ChatCompletionOverDatasetInput
+    ) -> AsyncIterator[ChatCompletionSubscriptionPayload]:
+        provider_key = input.model.provider_key
+        llm_client_class = PLAYGROUND_CLIENT_REGISTRY.get_client(provider_key, input.model.name)
+        if llm_client_class is None:
+            raise BadRequest(f"No LLM client registered for provider '{provider_key}'")
+
+        dataset_id = from_global_id_with_expected_type(input.dataset_id, Dataset.__name__)
+        version_id = (
+            from_global_id_with_expected_type(
+                global_id=input.dataset_version_id, expected_type_name=DatasetVersion.__name__
+            )
+            if input.dataset_version_id
+            else None
+        )
+        revision_ids = (
+            select(func.max(models.DatasetExampleRevision.id))
```

Comment:
I think this is now the 3rd place we have logic like this (gql queries, rest queries, here) and I think it *might* be worthwhile to start figuring out if this is worth extracting away

Reply:
That makes sense. Let's tackle in a follow-up.